Use a specific value for nodePort when the type of Service is NodePort

geoand opened the pull request:
The idea behind this is that subsequent updates / pushes to a cluster
using a NodePort Service (most likely to be used with Minikube)
will result in the same external URL being used.
This way users don't have to find the new URL each time they update / push.

geoand commented:
@iocanel I am not sure that the Decorator implementation I added is the most appropriate...

@Ladicek @maxandersen I would like to hear your thougths as well.

geoand commented:
@maxandersen brought up the point that this will fail if more than one replicas are deployed.

Max also raised the point that these kind of settings should be applied when the user is in "development mode", which unfortunately for the time being doesn't tied in to the Quarkus Dev Mode, since the Kubernetes extension requires the prod jar to be built.

My idea with this PR is to make the deployment / update experience for a user using Mininkube as friction-less as possible.
In my view when a user uses type `NodePort` for a service, it's pretty much guaranteed that they  in a sort of "developer mode", not production mode, therefore we should make it as easy as possible to deploy and update the application.
This also means that the user will probably be using a single replica (we can of course warn when there are more). 
Another idea I had was to not use a specific nodePort as the default, but to derive it from the application name - thus guaranteeing that it stays the same between application updates. Again this would only be done when `nodePort` is set for the `service-type` and would only be done when a single replica is used.

Thoughts?

iocanel reviewed (changes requested):
I am a bit skeptical about using a fixed node port.

What will happen if for any reason the port is already taken?
What will happen if I am using the `kubernetes` extension in multiple applications or modules?

So, I would like to do some small research for alternatives.

If no other alternative can be found, then I would add the following behavior:

1. Feature flag to turn this on and off (default value is debatable).
2. I would hash the service name and namespace and get the default node port (to prevent conflicts if no node port has been explicitly specified).
3. Deployer should display the node port to the user.

Ladicek commented on extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddNodePortDecorator.java:
```diff
@@ -0,0 +1,35 @@
+
+    public AddNodePortDecorator(String name, int nodePort) {
+        super(name);
+        if (nodePort < MIN_VALUE || nodePort > MAX_VALUE) {
```
This https://kubernetes.io/docs/concepts/services-networking/service/#nodeport says that the default range is 30000-32767, and apparently it's configurable.

geoand replied:
> I am a bit skeptical about using a fixed node port.
> 
> What will happen if for any reason the port is already taken?
> What will happen if I am using the `kubernetes` extension in multiple applications or modules?
> 
> So, I would like to do some small research for alternatives.
> 
> If no other alternative can be found, then I would add the following behavior:
> 
> 1. Feature flag to turn this on and off (default value is debatable).

Why would you need a feature flag? Isn't the use of `NodePort` along with `1` replica enough?

> 2. I would hash the service name and namespace and get the default node port (to prevent conflicts if no node port has been explicitly specified).

Yes, this is pretty much what I suggested above.

> 3. Deployer should display the node port to the user.

Good idea

geoand replied:
Ah, nice find! I need to remove these checks then.

Ladicek commented on extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesConfig.java:
```diff
@@ -107,6 +107,12 @@
+    /**
+     * The nodePort to set when serviceType is set to nodePort
+     */
+    @ConfigItem(defaultValue = "31987")
```
Is there a way to _not_ specify the concrete nodePort and let Kubernetes assign a port automatically? Perhaps allow the 0 value, which typically means "port should be auto-assigned"?

geoand replied:
When you don't set it, it gets assigned automatically. The problem is that it gets **re**-assigned to another value when you update the application and that's what I am trying to avoid.

Ladicek replied:
Sure. What I'm trying to say is: is there a way how to get back to the default behavior?

geoand replied:
Yeah, that essentially boils down to what @iocanel  is asking for :)

geoand commented:
> I am a bit skeptical about using a fixed node port.
> 
> What will happen if for any reason the port is already taken?
> What will happen if I am using the `kubernetes` extension in multiple applications or modules?
> 
> So, I would like to do some small research for alternatives.
> 
> If no other alternative can be found, then I would add the following behavior:
> 
> 1. Feature flag to turn this on and off (default value is debatable).

Why would you need a feature flag? Isn't the use of `NodePort` along with `1` replica enough?

> 2. I would hash the service name and namespace and get the default node port (to prevent conflicts if no node port has been explicitly specified).

Yes, this is pretty much what I suggested above.

> 3. Deployer should display the node port to the user.

Good idea

Ladicek commented:
I'd say there's too much assumptions in this, so there should be a way out. Other than that, I don't have an opinion, because I have no idea what NodePort services are good for.

geoand commented:
> I'd say there's too much assumptions in this, so there should be a way out. Other than that, I don't have an opinion, because I have no idea what NodePort services are good for.

AFAIK, they are only good for "local development" - meaning they are the easiest way to expose a service to the outside world.

iocanel commented:
> > I am a bit skeptical about using a fixed node port.
> > What will happen if for any reason the port is already taken?
> > What will happen if I am using the `kubernetes` extension in multiple applications or modules?
> > So, I would like to do some small research for alternatives.
> > If no other alternative can be found, then I would add the following behavior:
> > 
> > 1. Feature flag to turn this on and off (default value is debatable).
> 
> Why would you need a feature flag? Isn't the use of `NodePort` along with `1` replica enough?

Not really, the user should be able to select 1 replica, nodePort without using a fixed port.

> 
> > 1. I would hash the service name and namespace and get the default node port (to prevent conflicts if no node port has been explicitly specified).
> 
> Yes, this is pretty much what I suggested above.

I am guilty of posting without reading :-D

> 
> > 1. Deployer should display the node port to the user.
> 
> Good idea

iocanel commented:
If we get the deployer to display the node port to the user, we could make it somewhat smarter and not reapply the service if it hasn't actually changed (in the hope of the using the same port), so that the user doesn't have to change ports all the time.

geoand commented:
So if I add feature flag and nice logging messages, do we agree on this approach?

geoand commented:
> If we get the deployer to display the node port to the user, we could make it somewhat smarter and not reapply the service if it hasn't actually changed (in the hope of the using the same port), so that the user doesn't have to change ports all the time.

This seems more complicated to get right to me.

maxandersen commented:
The docs on nodeport gives the reasons what nodeport is useful for:

```
Using a NodePort gives you the freedom to set up your own load balancing solution, to configure environments that are not fully supported by Kubernetes, or even to just expose one or more nodes’ IPs directly.
```

We are kinda doing the latter here.

I still think the default behavior is best kept at "let kubernetes deal with it" but make it easy to activate this "dev-deployment" mode.

geoand commented:
It looks like having more than one replica, does not cause a problem with `nodePort`. However the exposed service only "targets" a single one of them, the other replicas are not accessible from outside the cluster.

geoand commented:
> The docs on nodeport gives the reasons what nodeport is useful for:
> 
> ```
> Using a NodePort gives you the freedom to set up your own load balancing solution, to configure environments that are not fully supported by Kubernetes, or even to just expose one or more nodes’ IPs directly.
> I still think the default behavior is best kept at "let kubernetes deal with it" but make it easy to activate this "dev-deployment" mode.

That seems to be the consencus here - if I have read all the comments correctly that is. 
I'll look into it.

maxandersen commented:
> It looks like having more than one replica, does not cause a problem with `nodePort`. However the exposed service only "targets" a single one of them, the other replicas are not accessible from outside the cluster.

are you sure ? I read the docs stating this: "you can specify a value in the nodePort field. The control plane will either allocate you that port or report that the API transaction failed. This means that you need to take care of possible port collisions yourself. You also have to use a valid port number, one that’s inside the range configured for NodePort use."

geoand commented:
> > It looks like having more than one replica, does not cause a problem with `nodePort`. However the exposed service only "targets" a single one of them, the other replicas are not accessible from outside the cluster.
> 
> are you sure ? I read the docs stating this: "you can specify a value in the nodePort field. The control plane will either allocate you that port or report that the API transaction failed. This means that you need to take care of possible port collisions yourself. You also have to use a valid port number, one that’s inside the range configured for NodePort use."

Yes, I just tried it. There was no failure, all pods came up and the service was exposed, but only one of the pods was receiving traffic.

iocanel commented:
> > > It looks like having more than one replica, does not cause a problem with `nodePort`. However the exposed service only "targets" a single one of them, the other replicas are not accessible from outside the cluster.
> > 
> > 
> > are you sure ? I read the docs stating this: "you can specify a value in the nodePort field. The control plane will either allocate you that port or report that the API transaction failed. This means that you need to take care of possible port collisions yourself. You also have to use a valid port number, one that’s inside the range configured for NodePort use."
> 
> Yes, I just tried it. There was no failure, all pods came up and the service was exposed, but only one of the pods was receiving traffic.

Which makes sense cause you are not load balancing.
This is a sideffect of using `NodePort` not a `fixed` `NodePort`.

geoand commented:
Here is my latest idea:

We don't use a flag per se, but instead have an enum configuration value named `quarkus.kubernetes.dev-mode` (or something like that) that behaves like so:

* OFF - No changes are made whatsoever
* ON - We enable `nodePort` and pick a proper port (and in the future perhaps we do more)
* INFER - We try to infer whether the cluster is Minikube (and whatever else makes sense) and if it is we make the same changes as if `ON` was set.

The default should be `OFF` most likely.

What do you think?

iocanel commented:
IMHO, it doesn't feel quite right.

For starters, we shouldn't `INFER`. Anything related to talking to an existing cluster should be off the table. In the same spirit, the same should apply to things like `minikube status` and the likes :-D

I also don't like the `dev-mode` thing, as it might be too confusing.
When the user specifies `NodPort` type, we could do the following:
1. Use the provided port.
2. Use a unique port based on name & namespace (if no port is specified).
3. Let kubernetes pick port if the user explicitly specifies 0 as the desired port value.

`3` allows us to disable the special feature without introducing weird-looking properties.

Now, since we can't query the environment for minikube / minishift cluster status etc, I am wondering if we could introduce `minikube` and `minishit` as individual deployment targets, with their own presets.

geoand commented:
> IMHO, it doesn't feel quite right.
> 
> For starters, we shouldn't `INFER`. Anything related to talking to an existing cluster should be off the table. In the same spirit, the same should apply to things like `minikube status` and the likes :-D
> 

I don't see why we shouldn't do this. It's purely a dev thing, so why not make developers lives easier?

> I also don't like the `dev-mode` thing, as it might be too confusing.
> When the user specifies `NodPort` type, we could do the following:
> 
> 1. Use the provided port.
> 2. Use a unique port based on name & namespace (if no port is specified).
> 3. Let kubernetes pick port if the user explicitly specifies 0 as the desired port value.
> 
> `3` allows us to disable the special feature without introducing weird-looking properties.
> 

Letting Kubernetes pick the port however disrupts the developers workflow by changing the URL that is used to access the application. That's what I am really trying to avoid here.

> Now, since we can't query the environment for minikube / minishift cluster status etc, I am wondering if we could introduce `minikube` and `minishit` as individual deployment targets, with their own presets.

Sure, but why do this if we can infer the environment?

davsclaus commented:
Having a really awesome and easy "kick the tires" experience for new users to quarkus and kubernetes running on their local developer laptop (especially for windows, mac users) is something that can really help quarkus and kick us ahead. 

Just think what Docker Desktop does for user experience and how to install and use it.

Minikube is reasonable easy to install - and having a great blog, and video how to do the code.quarkus.io -> zip -> docker mutlistage build -> kubectl apply -f -> runs on kubernetes with 4mb of memory usage.

And you would need the NodePort to call your deployed kubernetes.

And btw would also be good if quarkus could log on startup whether it runs in JVM or Native mode.

Ladicek commented:
If we're after good developer experience, what's the point of messing with NodePorts when you can `minikube addons enable ingress` and do everything properly?

geoand commented:
> If we're after good developer experience, what's the point of messing with NodePorts when you can `minikube addons enable ingress` and do everything properly?

Because `NodePort` is the easiest OOTB experience :)

Ladicek commented:
That I find hard to agree with.

Ladicek commented:
In any case, if we make nodePort configurable (I don't disagree with that!), and if we make the default value some specific number (I don't disagree with that either, if it's documented), then there needs to be a way how to get back to the default behavior (automatic assignment), because you just can't assume that NodePorts will only be used for the scenario you're thinking of right now.

iocanel commented:
> > IMHO, it doesn't feel quite right.
> > For starters, we shouldn't `INFER`. Anything related to talking to an existing cluster should be off the table. In the same spirit, the same should apply to things like `minikube status` and the likes :-D
> 
> I don't see why we shouldn't do this. It's purely a dev thing, so why not make developers lives easier?
> 
> > I also don't like the `dev-mode` thing, as it might be too confusing.
> > When the user specifies `NodPort` type, we could do the following:
> > 
> > 1. Use the provided port.
> > 2. Use a unique port based on name & namespace (if no port is specified).
> > 3. Let kubernetes pick port if the user explicitly specifies 0 as the desired port value.
> > 
> > `3` allows us to disable the special feature without introducing weird-looking properties.
> 
> Letting Kubernetes pick the port however disrupts the developers workflow by changing the URL that is used to access the application. That's what I am really trying to avoid here.
> 
> > Now, since we can't query the environment for minikube / minishift cluster status etc, I am wondering if we could introduce `minikube` and `minishit` as individual deployment targets, with their own presets.
> 
> Sure, but why do this if we can infer the environment?

IMHO resource generation should be `idempotent` ( should never rely on external things ).
Having generated resources change from machine to machine and from time to time based on what `minikube status` returns guaranteed to cause issues. We've tried in the past with FMP and it didn't end up well.

geoand commented:
@iocanel fair point, however this feature would be opt-in, not the default.

maxandersen commented:
> In any case, if we make nodePort configurable (I don't disagree with that!), and if we make the default value some specific number (I don't disagree with that either, if it's documented), then there needs to be a way how to get back to the default behavior (automatic assignment), because you just can't assume that NodePorts will only be used for the scenario you're thinking of right now.

this looks most sane to me; but you'll still somehow know to toggle from clusterip to nodeport, right ?

geoand commented:
> > In any case, if we make nodePort configurable (I don't disagree with that!), and if we make the default value some specific number (I don't disagree with that either, if it's documented), then there needs to be a way how to get back to the default behavior (automatic assignment), because you just can't assume that NodePorts will only be used for the scenario you're thinking of right now.
> 
> this looks most sane to me; but you'll still somehow know to toggle from clusterip to nodeport, right ?

Yes, unless we introduce what I was talking about: `quarkus.kubernetes.dev-mode`

Ladicek commented:
Reading `quarkus.kubernetes.dev-mode` immediately makes me ask: what's the relationship with Quarkus dev mode?

geoand commented:
Unfortunately at this point, none...

The Kubernetes extension requires the regular jar to be built, which we don't have in dev-mode.

No doubt the name I proposed sucks, but I am hoping that we can get past the name :P

Ladicek commented:
> this looks most sane to me; but you'll still somehow know to toggle from clusterip to nodeport, right ?

Can we make a better `quarkus.kubernetes.expose`? One that would take, instead of just `true`/`false`, an enum of `nodeport`/`loadbalancer`/`ingress` (and perhaps `none`). If it was configured to `ingress`, it would work as today's `true` (create an Ingress resource). If it was configured to `nodeport` or `loadbalancer`, no Ingress would be created, but the Service resource we generate would be of type NodePort / LoadBalancer. (The port number for NodePort would be configured using the config property proposed in this PR. There's a variety of options how to choose a default value; a hash of the GAV in the 30000-32767 range is probably a good idea. If the user does `quarkus.kubernetes.deploy`, we could/should even print the URL.)

geoand commented:
I am fine with @Ladicek's idea. 

I don't know if we can now the URL in all cases, but if we can it would be great.

geoand commented:
@iocanel @maxandersen @davsclaus WDYT about the enhanced expose idea?

davsclaus commented:
Just a thought I had yesterday - what if we generate a kubernetes-minikube yaml|json files that are adjusted for a single node locale development experience. Then users can use this file to deploy. Then it can use NodePort and the image pull thing.

1)
To generate this file, there can be a option to enable (find a good name)

quarkus.kubernetes.minikube=true

And also the maven plugin can have a configuration you can set via -D jvm system parameter and in the pom.xml <configuration> section.

or 

2)
maybe create a new kubernetes-minikube extension, so you just add this extension and the tooling adjusts the manifest generation for minikube user experience.

geoand commented:
@davsclaus I think @iocanel had a similar idea - essentially creating a "dev-mode" kubernetes file where various developer experience related changes are applied and that file would be used to deploy.

Like I told him as well, I am not at all opposed to the idea, if we agree on it, it's fine with me.

maxandersen commented:
Starting to warm up for the idea that @iocanel made to make minikube extension similar to openshift extension.

Allowing you to explicilty say what "kind" of deploy you are doing in which profile...

maxandersen commented:
still relelvant to have a "devmode" deploy for just openshift -one where images are not cached etc.  i don't think it make sense to have minikube be devmode version of openshift - its not  what users would end up doing in most cases anyway.

Ladicek commented:
Would it be better, instead of having `io.quarkus:quarkus-minikube`, have `quarkus.kubernetes.deployment-target=minikube` (in addition to the rest of values it currently supports0?

geoand commented:
Sounds reasonable to me.

iocanel commented:
So, the action items are the following:
- Enhance expose
- Introduce minikube resources (triggered by quarkus.kubernetes.deployment-target=minikube).
- When on minikube use expose=NodePort + George's idea.

I think that it starting to make a lot of sense.

geoand commented:
OK. I'll try and take them up later this week.

Ladicek commented:
I think there's some overlap between `quarkus.kubernetes.deployment-target=minikube` and `quarkus.kubernetes.expose=nodeport`. Enhanced `expose` probably has value on its own, but if we `deployment-target=minikube`, would we treat the `expose` config in some special way? (I don't know, something like: if `minikube`, then `expose=true` actually means `expose=nodeport`?) It's getting a bit more complex with an additional deployment target, which is kinda Kubernetes, but not totally...?

I think I'll let you guys figure it out somehow and complain when I see it :-)

maxandersen commented:
outlined plan sounds reasonable, as @Ladicek says - once we see it we'll better know if it works ;)
 
@iocanel for change of expose; wouldn't we want to have a separate property rather than changing true/false and break users existing configs ?

and would we put this minikube behavior in a separate extension similar to openshift/s2i ?

iocanel commented:
> outlined plan sounds reasonable, as @Ladicek says - once we see it we'll better know if it works ;)
> 
> @iocanel for change of expose; wouldn't we want to have a separate property rather than changing true/false and break users existing configs ?

Good point!
Maybe we could introduce `exposeStrategy` and deprecate `expose` so that we don't break compatibility moving forward.

> 
> and would we put this minikube behavior in a separate extension similar to openshift/s2i ?
Another module sounds good to me.

iocanel commented:
> I think there's some overlap between `quarkus.kubernetes.deployment-target=minikube` and `quarkus.kubernetes.expose=nodeport`. Enhanced `expose` probably has value on its own, but if we `deployment-target=minikube`, would we treat the `expose` config in some special way? (I don't know, something like: if `minikube`, then `expose=true` actually means `expose=nodeport`?) It's getting a bit more complex with an additional deployment target, which is kinda Kubernetes, but not totally...?
> 
> I think I'll let you guys figure it out somehow and complain when I see it :-)

Hmmm, there is definitely overlap. The way I see it is between `expose/exposeStrategy=NodePort` and `serviceType=NodePort`.

The `minikube` thing doesn't overlap as it acts as a feature flag for certain conversations we may wish to apply, for example:

- allow null docker registry (not mandatory in minikube, due to shared docker daemon).
- allow explicitly setting the NodePort.
etc.

So I guess we might want to rethink if we really need the enhanced `expose`.

geoand commented:
It seems to me that if we move to the `minikube` route, we don't need to enhance `expose` - at least not for now. We might indeed need it in the future.

Ladicek commented:
I can certainly see `expose` retaining the old `true`/`false` values as equivalents of `ingress`/`none` (respectively), to avoid breaking users.

But, as @geoand says, perhaps that isn't necessary. Dunno.

geoand commented:
Yeah we really need to see something in action and use it before deciding what's best. I'll try and get the `minikube` thing done this week since that's one point that everyone seems to agree on.

geoand committed:
Introduce a Minikube specific deployment target for Kubernetes

Ladicek commented:
I'll tell you what, I'm no longer that keen on `deployment-target=minikube`, because it brings a bunch of other hassles (`src/main/kubernetes/minikube.yaml`?). We should just generate 2 set of files for `deployment-target=kubernetes`, the ordinary `kubernetes.{json,yaml}`, and the new `minikube.{json,yaml}` :-)

maxandersen commented:
@Ladicek you want always kubernates and minikube generated even you don't use minikube ? 
how about crc.yaml and minishift.yaml or kubernetes-desktop.yaml ...

why is it we think minikube is the "chosen one" ?

Ladicek commented:
I don't mind `minishift.yaml` or `crc.yaml` or whatever, if there's a need for those. We're after user-friendliness, no? ;-)

Adding a new `deployment-target` has consequences. Should we also add `quarkus.minikube.*` config? Should we support `src/main/kubernetes/minikube.yaml`? Isn't it too much if we want to tweak a few things for the Minikube scenario. I wouldn't mind if the file was called `kubernetes-minikube.{json,yaml}`, except it would make tab-completion worse for people who want `kubernetes.{json,yaml}`.

geoand commented:
@Ladicek if you don't use the deployment target, how would you know which file to apply when time comes to deploy the application?

Ladicek commented:
You mean `quarkus.kubernetes.deploy`? Ah, good point...

OK, I'll stay silent until there's working code and then we can debate again :-)

geoand commented:
> You mean `quarkus.kubernetes.deploy`? Ah, good point... 

Exactly.

> OK, I'll stay silent until there's working code and then we can debate again :-)

I hope to have something concrete we can talk about later on today.

geoand commented:
PR has been updated.

When `quarkus.kubernetes.deployment-target` contains `minikube`, the `minikube.yaml` is generated. 
Furthermore when `quarkus.kubernetes.deploy=true` then `minikube.yaml` is used if the previous condition applies.

Currently `minikube.yaml` is the same as `kubernetes.yaml`, except that is always uses `IfNotPresent` for `ImagePullPolicy`, `NodePort` for `ServiceType` and always has a specific `nodePort` value (either set by the user or automatically generated in the `30000-31999` range based on the name of the deployment).

iocanel commented:
> PR has been updated.
> 
> When `quarkus.kubernetes.deployment-target` contains `minikube`, the `minikube.yaml` is generated.
> Furthermore when `quarkus.kubernetes.deploy=true` then `minikube.yaml` is used if the previous condition applies.

Up until now, the deployer was always picking the first target (if multiples are defined). I think its good to have a single approach to the subject.

geoand commented:
Problem is that the "first" one was random. I attempted to clean it up and make it a lot more predictable.

Ladicek commented on extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeployer.java:
```diff
@@ -65,23 +67,72 @@ public void deploy(KubernetesClientBuildItem kubernetesClient,
+                checkForMissingRegistry = false;
+            } else if (kubernetesDeploymentTargets.size() > 1) {
+                throw new IllegalStateException(
+                        "Multiple Kubernetes deployment targets present with no known priorities. Please select a single deployment target using the \"quarkus.kubernetes.deploymentTarget\" property.");
```
Is it `deployment-target` or `deploymentTarget`?

geoand replied:
It's obviously the latter :P

Ladicek commented on extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java:
```diff
@@ -118,6 +125,13 @@ public void checkKubernetes(BuildProducer<KubernetesDeploymentTargetBuildItem> d
     }
 
+    @BuildStep
+    public void checkMinishift(BuildProducer<KubernetesDeploymentTargetBuildItem> deploymentTargets) {
```
Minikube, not Minishift :-)

geoand replied:
Oh man.. I don't know how many times I made that mistake...

Ladicek commented on extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java:
```diff
@@ -405,12 +422,19 @@ private void applyConfig(Session session, Project project, String target, String
             Capabilities capabilities) {
+
+        // no need to change when we use Minikube only
+        if ((deploymentTargets.size() == 1) && MINIKUBE.equals(deploymentTargets.iterator().next())) {
```
`... && deploymentTargets.contains(MINIKUBE)` should work

geoand replied:
Right, it's left over from what I was doing in an earlier iteration

Ladicek commented:
The implementation looks fine to me. Whether this is what we wanna do, I don't know, but it's probably one of the better options :-)

geoand committed:
Sanitize the way quarkus.kubernetes.deployment-target works

Essentially now the first user supplied value is used to deploy.
When no user supplied value exists, then priorities are used to
determine which target should be deployed

geoand committed:
Add a minikube extension as convenience for deploying to Minikube

geoand commented:
I added two more things:

* A `minikube` extension that simply enables the `minikube` deployment-target
* The way the deployment target is chosen has been sanitized to take into account the user's explicit choice if it exists, or implicitly enabled deployment targets otherwise (based on a priority to each one that I added).

@iocanel can you please take a look?

geoand commented:
This also needs documentation updates, but I would like to make them after this (and potential follow ups) are in.

maxandersen commented:
> I don't mind `minishift.yaml` or `crc.yaml` or whatever, if there's a need for those. We're after user-friendliness, no? ;-)

Yes, and filling up your `target` folder with superflouous yaml ain't what I consider userfriendly :)

geoand commented:
The implementation now *does* generate an additional `minishift.yaml`.

That way the use of the deployment target is consistent - minishift is just another deployment target.

maxandersen commented:
> The implementation now _does_ generate an additional `minishift.yaml`.

but only when explicitly requested in deployment target, right ?

geoand commented:
Exactly.

Or when the minikube extension is added and no deployment target is specified by the user - the same way that the Openshift extension works.

Ladicek commented:
Well, I don't like the final experience here, because you need to rebuild your program if you want to deploy it to another cluster, but apparently adding YAML files to `target` is worse, so whatever.

geoand commented:
> Well, I don't like the final experience here, because you need to rebuild your program if you want to deploy it to another cluster, but apparently adding YAML files to `target` is worse, so whatever.

If you have a better proposal for handling that problem, I'm all ears.

maxandersen commented:
> but apparently adding YAML files to `target` is worse, so whatever.

it is not worse - but why is minikube the thing to treat as a special case ? as far as I can see unless you are deploying to production then the deployment descriptors will need to be slightly different towards all known kubernetes systems ... gcp, eks, minikube, openshift, crc, etc.

geoand commented:
It's not a special case per-se. We could easily reuse the same kind of solution developed for this PR to other flavors of Kubernetes.
Then the deployment target determines what gets deployed.
If there is no deployment target set by the user, then whichever implicitly enabled one has the highest priority, wins.

I don't see how it could be more straight forward than that

geoand commented:
When talking with @iocanel, he mentioned that the original intent of `deployment-target` was meant to honor the user provided value.
This PR now does do that, it just also adds a way to determine which `deployment-target ` to actually apply when the user hasn't specified one.

maxandersen commented:
> I don't see how it could be more straight forward than that

yes, I like the proposed solution. I just got the feeling @Ladicek feels otherwise and i'm curious what he suggest otherwise.

geoand commented:
> > I don't see how it could be more straight forward than that
> 
> yes, I like the proposed solution. I just got the feeling @Ladicek feels otherwise and i'm curious what he suggest otherwise.

Same here :)

Ladicek commented:
So, perhaps I misunderstood something, but I realized yesterday that the `quarkus.kubernetes.deployment-target` has 2 distinct purposes: select which [many] YAML files to generate, and select which [single] YAML file to deploy when `quarkus.kubernetes.deploy=true`. I think that's probably what's been bugging me.

If we really want to have `io.quarkus:quarkus-minikube`, then perhaps `quarkus.kubernetes.deployment-target` should never affect which [many] YAML files are generated (this is selected by the extensions that are included), and it is only used to select which [single] YAML file is deployed. (Alternatively, would it be possible to discover the kind of the cluster we are connected to, and select the [single] YAML file based on that info?)

geoand commented:
> So, perhaps I misunderstood something, but I realized yesterday that the `quarkus.kubernetes.deployment-target` config property has 2 distinct purposes: select which [many] YAML files to generate, and select which [single] YAML file to deploy when `quarkus.kubernetes.deploy=true`. I think that's probably what's been bugging me.
> 

That understanding is correct.

> If we really want to have `io.quarkus:quarkus-minikube`, then perhaps `quarkus.kubernetes.deployment-target` should never affect which [many] YAML files are generated (this is selected by the extensions that are included), and it is only used to select which [single] YAML file is deployed. (Alternatively, would it be possible to discover the kind of the cluster we are connected to, and select the [single] YAML file based on that info?)

So currently if the user adds the `quarkus-minikube` extension that I added in this PR, when the then when no `deployment-target` is selected, both `minikube.yaml` and `kubernetes.yaml` are geneated, but `minikube.yaml` is applied when `quarkus.kubernetes.deploy=true` is set.
If the user wants to do something different, he/she needs to explicitly set `deployment-target`.

I personally think this behavior is reasonable, but it definitely needs to be properly documented.

geoand commented:
> (Alternatively, would it be possible to discover the kind of the cluster we are connected to, and select the [single] YAML file based on that info?)

That was my initial thought as well. We could certainly add that as an enhancement

iocanel commented:
IIUC, this is perfectly aligned with what we currently have. So, this pull request just introduces a new deployment target, that the user can optionally enable. I like it!

iocanel commented on extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesConfig.java:
```diff
@@ -193,11 +199,12 @@
-    @ConfigItem(defaultValue = "kubernetes")
-    List<String> deploymentTarget;
+    @ConfigItem
+    Optional<List<String>> deploymentTarget;
```
I guess that `Optional` is introduced in order to be able to easily tell when this is a default value or a user provided one?

geoand replied:
Yes, exactly

iocanel reviewed (approved):
LGTM,

I would like though a minor additon: A testcase that shows that minikube resource is only added when explicitly specified.

geoand commented:
> LGTM,
> 
> I would like though a minor additon: A testcase that shows that minikube resource is only added when explicitly specified.

In `BasicKubernetesTest` we have:

```
        final Path kubernetesDir = prodModeTestResults.getBuildDir().resolve("kubernetes");
        assertThat(kubernetesDir)
                .isDirectoryContaining(p -> p.getFileName().endsWith("kubernetes.json"))
                .isDirectoryContaining(p -> p.getFileName().endsWith("kubernetes.yml"))
                .satisfies(p -> assertThat(p.toFile().listFiles()).hasSize(2));
```

which does test that when there is no `minikube` extension and that the minikube target isn't added, then no `minikube.yaml` is generated.

Did you have something more in mind @iocanel ?

maxandersen commented:
> Alternatively, would it be possible to discover the kind of the cluster we are connected to, and select the [single] YAML file based on that info?)

this is kinda what the original PR was doing which would make things slow and you would always have a remote lookup - which could fail if the cluster not available etc. 

Thus I like this issue much better where if you don't say anything it will pick the extension that are avialable to generate - but if any ambiguity you have to state which to use - either on command line or in application.properties to use as default.

geoand commented:
I think the myself, Ioannis and Max agree on the current approach, so unless @Ladicek objects very strongly, I believe the way forward is to merge this.

Ladicek commented:
You took it out of context. I only proposed discovering what kind of cluster is there _when stuff is being deployed to it_. Not when YAML files are generated. (If that would be possible, we could get rid of `quarkus.kubernetes.deployment-target` altogether.)

Sorry, I currently don't have time to respond to this properly.

Ladicek commented:
The current approach adds more inconsistency (new `quarkus-minikube` extension, new `quarkus.kubernetes.deployment-target=minikube` value, but not new `quarkus.minikube.*` configuration options?), which is bad IMO, but as I said, I unfortunately don't have time to respond properly :-(

geoand commented:
The idea is to create a new `minikube.yaml` based on defaults we think should apply for it. If a user doesn't want to use those, they can always just use the fully controllable `kubernetes.yaml`.

I don't think it's adding any inconsistency, it's just adding more options. But like I said, we need to move forward one way or another, so if you don't like it (which is perfectly understandable and acceptable), then we really need a proposal for an alternative.

Ladicek commented:
More options that are not orthogonal = inconsistency.

(My proposal? Determine which YAML files to generate solely based on which `quarkus-{kubernetes,minikube,openshift}` extensions are present. Ignore `quarkus.kubernetes.deployment-target` for that process. If possible to detect which cluster we're talking to, determine which YAML file to deploy based on that and get rid of `quarkus.kubernetes.deployment-target` completely; otherwise determine which YAML file to deploy based on `quarkus.kubernetes.deployment-target`.)

At this point, I cave in. I officially retract everything I wrote in this issue, please treat it as if I never commented. Sorry, I don't have the required time these days.

geoand commented:
@Ladicek thanks for the input.

The problem I see with your proposal is that for the example the `minikube` extensions always implies the `kubernetes` extension, so the only way to disambiguate when deploying would be to try to determine the target cluster - which I am not entirely opposed to, but I don't think should be the default.

At this point I'll leave it up to the man in change to make the call, @maxandersen :)

maxandersen commented:
I know you said you want to ignore but I think you made some interesting points @Ladicek   I'm going to ask in hope to better understand your concerns.

"not orthogonal = inconsistency" - is that the issue that there are (currently) no minishift.* options to tweak ? if yes, then this approach does not exclude this from happening/getting added if we get some usecases where relying on the kubernetes config is not enough for minikube variation.

About ignoring deployment target for generating files - maybe I'm being naive here but what would that give you if only the specific files will be used when deploying ? I'm missing some usecase where you think its valuable to have this behavior.

Ladicek commented:
I think what I'm trying to say is that this is getting progressively harder to understand. `minikube` now looks very much like `kubernetes` or `openshift` (or `knative`? Never played with that one), but it isn't. Instead of adding special cases to cater for more usecases, we should strive to generalize. My suggestions were not well thought through in the last few days, so I'm not going to add more :-)

geoand commented:
I rebases the PR onto master because I wanted to make sure that nothing from https://github.com/quarkusio/quarkus/pull/8894 would interfere with this one.

maxandersen commented:
> I think what I'm trying to say is that this is getting progressively harder to understand. `minikube` now looks very much like `kubernetes` or `openshift` (or `knative`? Never played with that one), but it isn't. 

I'm a bit puzzled by your comment here - minikube, kubernetes and openshift looks  much  a like since they *are* alike. 

`kubernetes` is the plain one, we add `minkube` deployment target to tweak defaults in how you would use it out-of-box but we don't add additional `minikube.*` properties as we believe (thus far) that minikube is as close as it gets to `kubernetes`.

for `openshift` we do the same, have an extension to tweak the default choices for a better out of box experience and since openshift actually have openshift specific settings we have those as properties you can tweak.

I feel the configuration and behaviors we allow to set here are  as close to reality as we can get.

> Instead of adding special cases to cater for more usecases, we should strive to generalize. My suggestions were not well thought through in the last few days, so I'm not going to add more :-)

I agree we should generalize where we can - I'm struggling to see where we can generalize more in this case. if you have some ideas i'm all ears on that!

on this issue I say we should LGTM as it fits into the current approach of deployment targets thus adding it does not "hurt" and if we find ways to generalize more it should be doable with 4 deployment target options as it would be with the current 3.

geoand commented:
I still need to follow up with some documentation
